csrc/alias_analysis.cpp

```diff
@@ -170,15 +174,108 @@ std::optional<Layout> mapInLayoutToOutRoot(
 }
 
 namespace {
+// Given a TV, returns its layout with repsect to the logical domain. When
+// `allocation` is a split of `logical`, walks backwards from `allocation` to
+// `logical` to find a permutation of `logical` that satisfies the order in
+// `allocation`. The returned contiguity is computed according to
+// splitContiguity and mergeContiguity.
+//
+// Example:
+//   input TV:
+//     logical: [b, s, h]
+//     allocation: [s, d, h/d, b]
+//     contiguity: [t, t, f, t]
+//   output layout:
+//     allocation: [s, h, b]
+//     contiguity: [t, f, t]
+//
+// I chose to canonicalize layouts to logical so I didn't have to change the
+// main bulk of alias analysis which was written for single GPU.
+std::optional<Layout> canonicalizeLayout(const TensorView* tv) {
+  const std::vector<IterDomain*>& logical = tv->getLogicalDomain();
+  const std::vector<IterDomain*>& allocation = tv->getMaybeAllocationDomain();
+  LinkedHashMap<IterDomain*, std::optional<bool>> allocation_to_contiguity;
+  for (auto&& [alloc_id, contiguity] : zip(allocation, tv->getContiguity())) {
+    allocation_to_contiguity.pushBack(alloc_id, contiguity);
+  }
+
+  for (Expr* transform : DependencyCheck::getAllExprsBetween(
+                             {logical.begin(), logical.end()},
+                             {allocation.begin(), allocation.end()}) |
+           std::views::reverse) {
+    if (auto* split = dynamic_cast<Split*>(transform)) {
+      // When split->outer() is parallelized and split->inner() is serial, we
+      // remove split->outer() regardless of its position and replace
+      // split->inner() with split->in(). This way, even when split->outer() is
+      // not adjacent to split->inner() (e.g. when it's outermost), we can
+      // still undo the split.
+      //
+      // Several other cases that I haven't implemented for simplicity.
+      //
+      // When split->outer() is serial and split->inner() is parallelized, we
+      // could remove split->inner() and replace split->outer() with
+      // split->in() regardless of split->inner()'s position.
+      //
+      // When split->outer() and split->inner() are both parallelized, we could
+      // replace either of them with split->in() and remove the other.
+      NVF_ERROR(!split->inner()->isParallelized());
+
+      const auto [outer_contiguity, next_i] =
+          allocation_to_contiguity.erase(split->outer());
+      if (!split->outer()->isParallelized()) {
+        // Don't check adjacency if split->outer() is parallelized.
```

Comment:
nit: since we are in the true block.
```suggestion
        // Check adjacency if split->outer() is not parallelized.
```